npeps/npep-187-ports-and-protocols.md

```diff
@@ -0,0 +1,408 @@
+- exactly one of `number` or `range` must be set in `port`
+
+### Option2: Join by protocol
+
+```yaml
+protocols:
+  tcp:
+    - port:
+        number: 8080
+    - port:
+        range: 
+          start: 8080
+          end: 9090
+  udp:
+    - port:
+        number: 8080
+    - port:
+        number: 9090
+  namedPort:
+    - http
+    - monitoring
+  icmp:
+    - type: 7
+      code: 3
+```
+
+In this case it makes more sense to also flatten the `port` match, because for TCP/UDP/SCTP protocols, port is the only option and is always required:
+
+```yaml
+protocols:
+  tcp:
+    - port: 8080
+    - portRange: 
+        start: 8080
+        end: 9090
+  udp:
+    - port: 8080
+    - port: 9090
+  namedPort:
+    - http
+    - monitoring
+  icmp:
+    - type: 7
+      code: 3
+```
+
+with the validation like so:
+- empty `protocols` list is not allowed
+- at least 1 field of `protocol` must be set
+- each protocol list must not be empty
+- exactly one of `port` or `portRange` must be set for TCP/UDP/SCTP elements
+
+### Option3: Support older clients
+
+Supporting older clients that may not know about the newer API fields is an important consideration for APIs where
+evolution is highly likely. The idea is to make it possible for older clients to understand/report that a
+given CNP has fields they don't understand and "fail closed".
+
+Suppose one day we add GRE support. The API understands an object with gre specified but we have an 
+older client which does not. The client will deserialize the gre field and discard it. 
+One approach to defending that is to elevate the list, like:
+
+```go
+type Protocols struct {
+// exactly one should be set, if you see this empty it's something you don't know how to handle
+    TCP *TCPProtocol
+    UDP *UDPProtocol
+    GRE *GREProtocol
+}
+```
+
+Then rather than the parent having a single `Protocols Protocols` it would have a list: `Protocols []Protocols`. Rendering as YAML like:
+
+```yaml
+protocols
+- tcp:
+    port: 53
+- udp:
+    port: 53
+- gre:
+    somethingHere: true
+```
+and an older client would see:
+```yaml
+protocols
+- tcp:
+    port: 53
+- udp:
+    port: 53
+- {}  // <- something is obviously wrong!
+```
+
+Our CRD has only 1 implementation per cluster and both are usually installed and updated together by the cluster admin
+(or a CRD itself may be installed by the implementation).
+So the implementation should know all of the fields that exist in the CRD (though if they're installing the experimental CRD, 
+they might not implement every field in the CRD, but in that case they are responsible for noticing that one of the fields 
+they don't implement is set).
+
+We consider a situation where an older implementation is installed with a newer CRD unlikely/misconfiguration.
+The existing convention on how to handle unknown fields in rules is:
+- If there's something you don't recognize in an `action: Accept` rule, then you must assume that the rule does not match anything 
```

Comment:
Does a generated client even know if there are unmatched fields? This would require using Unstructured Client for the data structure?